stats/stats-server/src/update_service.rs

```diff
@@ -27,74 +31,98 @@ fn time_till_next_call(schedule: &Schedule) -> std::time::Duration {
 impl UpdateService {
     pub async fn new(
         db: Arc<DatabaseConnection>,
-        blockscout: Arc<DatabaseConnection>,
+        blockscout_db: Arc<DatabaseConnection>,
         charts: Arc<RuntimeSetup>,
+        status_listener: Option<IndexingStatusListener>,
     ) -> Result<Self, DbErr> {
         Ok(Self {
             db,
-            blockscout,
+            blockscout_db,
             charts,
+            status_listener,
         })
     }
 
-    /// Perform initial update and run the service in infinite loop.
-    /// Terminates dependant threads if one fails.
-    pub async fn force_async_update_and_run(
+    /// The main function of the service.
+    ///
+    /// Run the service in infinite loop.
+    /// Terminates dependant threads if enough fail.
+    pub async fn run(
         self: Arc<Self>,
-        concurrent_tasks: usize,
+        concurrent_initial_tasks: usize,
         default_schedule: Schedule,
         force_update_on_start: Option<bool>,
     ) {
-        let semaphore = Arc::new(tokio::sync::Semaphore::new(concurrent_tasks));
-        let (tasks, mut updaters) = self
-            .charts
-            .update_groups
-            .values()
+        let initial_update_semaphore: Arc<tokio::sync::Semaphore> =
+            Arc::new(tokio::sync::Semaphore::new(concurrent_initial_tasks));
+        let groups = self.charts.update_groups.values();
+        let init_update_tracker = InitialUpdateTracker::new(groups.len() as u64);
+        let mut group_update_jobs: FuturesUnordered<_> = groups
             .map(|group| {
```

Comment:
nit: this function is rather big, it would be great to move complicated logic and show only business logic, like

```rust
.map(|group| {
    wait_until_status_ok(group)
    initial_group_update(group)
    run_infinite_update_loop(group)
})
```